hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotCreateRequest.java

```diff
@@ -0,0 +1,175 @@
+import static org.apache.hadoop.ozone.om.lock.OzoneManagerLock.Resource.BUCKET_LOCK;
+
+
+/**
+ * Handles CreateSnapshot Request.
+ */
+public class OMSnapshotCreateRequest extends OMClientRequest {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMSnapshotCreateRequest.class);
+
+  private final String snapshotPath;
+  private final String volumeName;
+  private final String bucketName;
+  private final String snapshotName;
+  private final SnapshotInfo snapshotInfo;
+
+  public OMSnapshotCreateRequest(OMRequest omRequest) {
+    super(omRequest);
+    CreateSnapshotRequest createSnapshotRequest = omRequest
+        .getCreateSnapshotRequest();
+    volumeName = createSnapshotRequest.getVolumeName();
+    bucketName = createSnapshotRequest.getBucketName();
+    String possibleName = createSnapshotRequest.getSnapshotName();
+    snapshotInfo =
+        SnapshotInfo.newInstance(volumeName, bucketName, possibleName);
+    snapshotName = snapshotInfo.getName();
+    snapshotPath = snapshotInfo.getSnapshotPath();
+  }
+
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+    final OMRequest omRequest = super.preExecute(ozoneManager);
+    // Verify name
+    OmUtils.validateSnapshotName(snapshotName);
+
+    UserGroupInformation ugi = createUGI();
+    String bucketOwner = ozoneManager.getBucketOwner(volumeName, bucketName,
+        IAccessAuthorizer.ACLType.READ, OzoneObj.ResourceType.BUCKET);
+    if (!ozoneManager.isAdmin(ugi) &&
+        !ozoneManager.isOwner(ugi, bucketOwner)) {
+      throw new OMException(
+          "Only bucket owners/admins can create snapshots",
+          OMException.ResultCodes.PERMISSION_DENIED);
+    }
+    return omRequest;
+  }
+  
+  @Override
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long transactionLogIndex,
+      OzoneManagerDoubleBufferHelper ozoneManagerDoubleBufferHelper) {
+
+    OMMetrics omMetrics = ozoneManager.getMetrics();
+    omMetrics.incNumSnapshotCreates();
+
+    boolean acquiredBucketLock = false, acquiredSnapshotLock = false;
+    IOException exception = null;
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+    
+    OMResponse.Builder omResponse = OmResponseUtil.getOMResponseBuilder(
+        getOmRequest());
+    OMClientResponse omClientResponse = null;
+    AuditLogger auditLogger = ozoneManager.getAuditLogger();
+
+    OzoneManagerProtocolProtos.UserInfo userInfo = getOmRequest().getUserInfo();
+    String key = snapshotInfo.getTableKey();
+    try {
+      // Lock bucket so it doesn't
+      //  get deleted while creating snapshot
+      acquiredBucketLock =
+          omMetadataManager.getLock().acquireReadLock(BUCKET_LOCK,
+              volumeName, bucketName);
+
+      acquiredSnapshotLock =
+          omMetadataManager.getLock().acquireWriteLock(BUCKET_LOCK,
+              volumeName, snapshotInfo.getSnapshotLockResourceName());
```

Comment:
I guess it would be better to create a new category of lock called SNAPSHOT_LOCK here.